### ISSUE TITLE
EI_EXPOSE_REP false positive

### ISSUE THREAD

**Issue by xenoterracide** [spotbugs/spotbugs#2552]
```java
  private final List<List<?>> grid;

  List<? extends List<?>> grid() {
    return grid.stream().map(Collections::unmodifiableList).toList();
  }
```

This creates an immutable copy, and so shouldn't trigger this rule. In fact I tried several variations, none of them didn't trigger this rule. I'm not actually certain what variation of a copy is expected.

https://github.com/xenoterracide/ai-wh40k/blob/71d80c5da6f36f9794ab2058424c25908f25f252/module/game/src/main/java/com/xenoterracide/ai/wh40k/game/GameMap.java

**Comment by gtoison** [comment]
The contract for `stream.toList()` is indeed to return an unmodifiable list, on the other hand `stream.collect(Collectors.toList())` explicitely says that there's no guarantee on what kind of list is returned.

**Comment by gtoison** [comment]
Sorry that wasn't super clear, EI_EXPOSE_REP is about not exposing the internal state of the object. 
This can be achieved by returning an immutable object or by returning a copy. It is not super clear to me if this immutability is supposed to be deep or shallow. Does it matter if the immutable (or the copy) shares some references with the initial object?

**Comment by xenoterracide** [comment]
... If you actually look at the linked code, which is all the code, it's not a sample repo. There is no object. The list of lists only contains nulls. "?" By itself would require casting to access and mutate something. So everything here is already immutable or a copy, actually it's both.

**Comment by gtoison** [comment]
Although it is relatively clear that the returned structure is an immutable copy, this is not easily checked by SpotBugs.
The static type is `List<List<>>` which is mutable, however the dynamic type is immutable.

I'm not sure how SpotBugs could analyse this, appart maybe adding a special case for `stream.toList()`, because it is immutable by contract.

I'll let others chime in if they have better ideas

**Comment by xenoterracide** [comment]
I suspect there are a couple of underlying list types in the JVM that you could look for. It's of course really sad that the standard collections contracts don't have immutable variants like they do in kotlin.

**Comment by gtoison** [comment]
I tried reproducing the issue in #2571 but the test passes, so it looks like the false positive was solved earlier
